### DISCUSSION TITLE
Safety: Upcoming change in Watchdog

### DISCUSSION THREAD

**Discussion by Peter-van-Tol** [Peter-van-Tol/LiteX-CNC#62] [Announcements]
There are some upcoming changes in the watchdog, for safety reasons:
- bug resolved in the driver. The signal when the watchdog has bitten was not relayed to the HAL.
- if the watchdog has bit and the `has_bitten` bit is `True`, the user can reset it to `False` to resume operation.
- added an physical `enable_out` pin to the watchdog. Which pin it has to be, can be chosen in the configuration (see example below). I use this pin with the HUB75HAT project. When the board is running `enable_out` is `HIGH`. When the watchdog times out (either due to communication disruption or LinuxCNC stops sending signals) the pin becomes `LOW`. On the HUB75HAT this means that all input and output will go in high-Z mode. This means the watchdog safety is completely in hardware and does not depend on LinuxCNC or the communication.

Example configuration (default for HUB75HAT):
```json
{
    "board_name": "EMCO5_SPI",
    "board_type": "HUB75HAT v8.0",
    "clock_frequency": 50000000,
    "connection": {
        "connection_type": "spi",
            "mosi": "spi:0",
            "miso": "spi:1",
            "clk":  "spi:2",
            "cs_n": "spi:3"
    },
    "watchdog": {
        "pin":"ena:0"
    },
    "modules": [
        ...
    ]
}
```

Two open actions before I commit this change are:
- making the physical `enable_out` pin optional;
- create a working example where the watchdog is part of the E-stop latch circuit. This requires some additional components, to drive the `has_bitten` pin low before enabling the E-stop chain.

**Comment by GTCLive** [comment]
Nice! yet important.

**Reply by Peter-van-Tol** [reply]
Thanks for your feedback. 

How about adding E-stop also to the watchdog, i.e.:
```json
    "watchdog": {
        "enable_out":"ena:0"
        "triggers": [
            {"pin": "j1:1", "trigger": "LOW"},
            {"pin": "j1:2", "trigger": "LOW"},
            ....
        ] 
    },
```

In this example one could tie for example ESTOP and the ALARM from stepgen directly to the watchdog. This ensures the machine will be disconnected when one of these occurs, even when the communication fails. It is up to the user to choose which signals to route through the watchdog, this depends on the machine and its configuration. But I would really like my external E-Stop to stop it directly.

**Comment by Peter-van-Tol** [comment]
The change (excluding the external triggers) has now been merged with `main`. It will be included in the next release.

**Comment by Peter-van-Tol** [comment]
Release is now being prepared. All images for the documentation are being re-build as we speak. Hopefully version 1.2 can be released tonight.
